pypesto/objective/history.py

```diff
@@ -135,7 +181,67 @@ def _update_counts(self, sensi_orders, mode):
             if 1 in sensi_orders:
                 self.n_sres += 1
 
-    def _update_trace(self, x, sensi_orders, mode, result):
+    def _fval2chi2_offset(self, x: np.ndarray, chi2: float):
+        """
+        Initialize the trace.
+        """
+        if self.fval2chi2_offset is None:
+            if self.obj is not None:
+                self.fval2chi2_offset = self.obj(x, (0,))[FVAL] - chi2
+            else:
+                self.fval2chi2_offset = 0.0
+
+        return self.fval2chi2_offset
+
+    def _init_trace(self, x: np.ndarray):
+        """
+        Initialize the trace.
+        """
+        if self.x_names is None:
+            self.x_names = [f'x{i}' for i, _ in enumerate(x)]
+
+        columns: List[Tuple] = [
+            (c, float('nan')) for c in [
+                'time', 'n_fval', 'n_grad', 'n_hess', 'n_res', 'n_sres',
+                'fval', 'chi2', 'res', 'sres', 'hess',
+            ]
+        ]
+
+        for var in ['x', 'grad', 'schi2']:
+            if var == 'x' or self.options[f'trace_record_{var}']:
+                columns.extend([
```

Comment:
what was the reason for having one field per entry, and not storing all in one field?

Reply:
storing arrays as string is not very robust and does not always roundtrip through saving as csv (see acrobatics with `ndarray2string_full` to ensure full numeric precision and completeness of results)